src/utils/net/download.rs

```diff
@@ -54,149 +53,125 @@ impl<R: AsyncRead + Unpin> AsyncRead for FetchProgress<R> {
     }
 }
 
-type DownloadStream = IntoAsyncRead<MapErr<hyper::Body, fn(hyper::Error) -> futures::io::Error>>;
+impl AsyncBufRead for FetchProgress {
+    fn poll_fill_buf(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<std::io::Result<&[u8]>> {
+        Pin::new(&mut Pin::get_mut(self).inner).poll_fill_buf(cx)
+    }
 
-impl FetchProgress<DownloadStream> {
-    pub async fn fetch_from_url(url: &Url) -> anyhow::Result<Self> {
-        let (inner, progress_bar) = fetch_stream_from_url(url).await?;
-        Ok(FetchProgress {
-            inner,
-            progress_bar,
-        })
+    fn consume(mut self: Pin<&mut Self>, amt: usize) {
+        Pin::new(&mut self.inner).consume(amt);
+        self.progress_bar.add(amt as u64);
     }
 }
 
-impl FetchProgress<ZstdDecoder<DownloadStream>> {
-    pub async fn fetch_zstd_compressed_from_url(url: &Url) -> anyhow::Result<Self> {
-        let (inner, progress_bar) = fetch_stream_from_url(url).await?;
-        let inner = ZstdDecoder::new(inner);
-        Ok(FetchProgress {
-            inner,
+/// FileReader facilitates file streaming, whether local or remote.
+pub struct FileReader {}
+
+impl FileReader {
+    /// Returns [`FetchProgress`]
+    ///
+    /// Auto-detects whether or not the file has to be streamed from a remote location, and
+    /// whether or not it's zstd compressed.
+    ///
+    /// # Arguments
+    ///
+    /// * `path` - Snapshot location, could be either remote or local
+    pub async fn read(
+        path: &str,
+    ) -> anyhow::Result<Either<ZstdDecoder<FetchProgress>, FetchProgress>> {
+        let is_remote_file: bool = path.starts_with("http://") || path.starts_with("https://");
```

Comment:
Would this work?
`let is_remote_file = Url::parse(path).is_ok()`

Reply:
Yeah, I think it'd be better. I kept this in order to be consistent with the initial implementation, but there's no reason why we couldn't accept more protocols.